packages/contracts-bedrock/test/L1/DataAvailabilityChallenge.t.sol

```diff
@@ -0,0 +1,504 @@
+pragma solidity 0.8.15;
+
+import {
+    DataAvailabilityChallenge,
+    ChallengeStatus,
+    Challenge,
+    CommitmentType,
+    computeCommitmentKeccak256
+} from "src/L1/DataAvailabilityChallenge.sol";
+import { Proxy } from "src/universal/Proxy.sol";
+import { CommonTest } from "test/setup/CommonTest.sol";
+
+contract DataAvailabilityChallengeTest is CommonTest {
+    DataAvailabilityChallenge public dac;
+
+    function setUp() public virtual override {
+        super.enablePlasma();
+        super.setUp();
+        dac = DataAvailabilityChallenge(deploy.mustGetAddress("DataAvailabilityChallengeProxy"));
+    }
+
+    function testDeposit() public {
+        assertEq(dac.balances(address(this)), 0);
+        dac.deposit{ value: 1000 }();
+        assertEq(dac.balances(address(this)), 1000);
+    }
+
+    function testReceive() public {
+        assertEq(dac.balances(address(this)), 0);
+        (bool success,) = payable(address(dac)).call{ value: 1000 }("");
+        assertTrue(success);
+        assertEq(dac.balances(address(this)), 1000);
+    }
+
+    function testWithdraw(address sender, uint256 amount) public {
+        assumePayable(sender);
+        assumeNotPrecompile(sender);
+        vm.assume(sender != address(dac));
+        vm.assume(sender.balance == 0);
+        vm.deal(sender, amount);
+
+        vm.prank(sender);
+        dac.deposit{ value: amount }();
+
+        assertEq(dac.balances(sender), amount);
+        assertEq(sender.balance, 0);
+
+        vm.prank(sender);
+        dac.withdraw();
+
+        assertEq(dac.balances(sender), 0);
+        assertEq(sender.balance, amount);
+    }
+
+    function testChallengeSuccess(address challenger, uint256 challengedBlockNumber, bytes calldata preImage) public {
+        bytes memory challengedCommitment = computeCommitmentKeccak256(preImage);
+
+        // Assume the challenger is not the 0 address
+        vm.assume(challenger != address(0));
+
+        // Assume the block number is not close to the max uint256 value
+        vm.assume(challengedBlockNumber < type(uint256).max - dac.challengeWindow() - dac.resolveWindow());
+        uint256 requiredBond = dac.bondSize();
+
+        // Move to a block after the challenged block
+        vm.roll(challengedBlockNumber + 1);
+
+        // Deposit the required bond
+        vm.deal(challenger, requiredBond);
+        vm.prank(challenger);
+        dac.deposit{ value: requiredBond }();
+
+        // Expect the challenge status to be uninitialized
+        assertEq(
+            uint8(dac.getChallengeStatus(challengedBlockNumber, challengedCommitment)),
+            uint8(ChallengeStatus.Uninitialized)
+        );
+
+        // Challenge a (blockNumber,hash) tuple
+        vm.prank(challenger);
+        dac.challenge(challengedBlockNumber, challengedCommitment);
+
+        // Challenge should have been created
+        (address _challenger, uint256 _lockedBond, uint256 _startBlock, uint256 _resolvedBlock) =
```

Comment:
I think its less error prone to return the struct in memory rather than 4 stack items but we are not necessarily consistent with this. solc will autogen bytecode that returns stack items like this, the way to get the struct in memory would be to make the `challenges` mapping internal and then make a public function that explicitly sets its return type to the struct. not super opinionated altho i do think it makes it less error prone because you don't need to use parens and know the order of the return values